oauth2client/flask.py

```diff
@@ -0,0 +1,299 @@
+
+        if resp.status != 200:
+            current_app.logger.error(
+                "Error while obtaining user profile: %s" % resp)
+            return None
+
+        profile = json.loads(content)
+        return profile
+
+    def _make_flow(self):
+        # Generate a CSRF token to prevent malicious requests.
+        state = ''.join(random.choice(string.ascii_uppercase + string.digits)
+                        for x in range(32))
+
+        session['oauth2state'] = state
+
+        return OAuth2WebServerFlow(
+            client_id=self.client_id,
+            client_secret=self.client_secret,
+            scope=self.scopes,  # TODO
+            state=state,
+            approval_prompt='force',  # TODO
+            redirect_uri=url_for("oauth2.callback", _external=True))
+
+    def _create_blueprint(self):
+        bp = Blueprint('oauth2', __name__)
+        bp.add_url_rule('/oauth2authorize', 'authorize', self.authorize_view)
+        bp.add_url_rule('/oauth2logout', 'logout', self.logout_view)
+        bp.add_url_rule('/oauth2callback', 'callback', self.callback_view)
+
+        return bp
+
+    def authorize_view(self):
+        return_url = request.args.get('return_url')
+        if not return_url:
+            return_url = request.referrer or '/'
+        session['oauth2return'] = return_url
+
+        flow = self._make_flow()
+        auth_url = flow.step1_get_authorize_url()
+        return redirect(auth_url)
+
+    def logout_view(self):
+        session.pop('oauth2credentials', None)
+        session.pop('oauth2profile', None)
+        session.pop('oauth2userid', None)
+
+        if self.logout_callback:
+            self.logout_callback()
+
+        return redirect(request.referrer or '/')
+
+    def callback_view(self):
+        # Check the CSRF token.
+        if request.args.get('state', '') != session.pop('oauth2state', None):
+            return 'Invalid request state', 400
+
+        flow = self._make_flow()
+
+        # Exchange the auth code for credentials.
+        try:
+            credentials = flow.step2_exchange(request.args['code'])
+        except FlowExchangeError as e:
+            current_app.logger.exception(e)
+            return e.message, 400
+
+        # Save the credentials to the session.
+        session['oauth2credentials'] = credentials.to_json()
+
+        # Get the user's profile from Google and save it to the session.
+        if 'profile' in self.scopes:
+            session['oauth2profile'] = self._request_user_info()
+
+        if self.authorize_callback:
+            self.authorize_callback()
+
+        return redirect(session.pop('oauth2return', '/'))
+
+    @property
+    def credentials(self):
+        ctx = _app_ctx_stack.top
+
+        if not hasattr(ctx, 'oauth2_credentials'):
+            serialized = session.get('oauth2credentials')
+
+            if not serialized:
+                return None
+
+            ctx.oauth2_credentials = Credentials.new_from_json(serialized)
```

Comment:
I was saying you should use `OAuth2Credentials.new_from_json` but doing a `git grep new_from_json` on the `oauth2client` source it doesn't seem to be used much.